aptos-move/move-examples/marketplace/sources/listing.move

```diff
@@ -0,0 +1,263 @@
+        let listing_addr = object::object_address(&object);
+        let Listing {
+            object,
+            seller,
+            fee_schedule,
+            start_time: _,
+            delete_ref,
+            extend_ref,
+        } = move_from<Listing>(listing_addr);
+
+        let obj_signer = object::generate_signer_for_extending(&extend_ref);
+        if (exists<TokenV1Container>(object::object_address(&object))) {
+            extract_or_transfer_tokenv1(recipient, object::convert(object));
+        } else {
+            object::transfer(&obj_signer, object, recipient);
+        };
+        object::delete(delete_ref);
+
+        (seller, fee_schedule)
+    }
+
+    public(friend) fun assert_started(object: &Object<Listing>): address acquires Listing {
+        let listing_addr = object::object_address(object);
+        assert!(exists<Listing>(listing_addr), error::not_found(ENO_LISTING));
+
+        let listing = borrow_global<Listing>(listing_addr);
+        let now = timestamp::now_seconds();
+        assert!(listing.start_time <= now, error::invalid_state(ELISTING_NOT_STARTED));
+        listing_addr
+    }
+
+    // View
+
+    #[view]
+    public fun seller(object: Object<Listing>): address acquires Listing {
+        let listing = borrow_listing(object);
+        listing.seller
+    }
+
+    #[view]
+    public fun listed_object(object: Object<Listing>): Object<ObjectCore> acquires Listing {
+        let listing = borrow_listing(object);
+        listing.object
+    }
+
+    #[view]
+    public fun fee_schedule(object: Object<Listing>): Object<FeeSchedule> acquires Listing {
+        let listing = borrow_listing(object);
+        listing.fee_schedule
+    }
+
+    #[view]
+    /// Compute the royalty either from the internal TokenV1, TokenV2 if it exists, or return
+    /// no royalty.
+    public fun compute_royalty(
+        object: Object<Listing>,
+        amount: u64,
+    ): (address, u64) acquires Listing, TokenV1Container {
+        let listing = borrow_listing(object);
+        let obj_addr = object::object_address(&listing.object);
+        if (exists<TokenV1Container>(obj_addr)) {
+            let token_container = borrow_global<TokenV1Container>(obj_addr);
+            let token_id = tokenv1::get_token_id(&token_container.token);
+            let royalty = tokenv1::get_royalty(token_id);
+
+            let payee_address = tokenv1::get_royalty_payee(&royalty);
+            let numerator = tokenv1::get_royalty_numerator(&royalty);
+            let denominator = tokenv1::get_royalty_denominator(&royalty);
+
+            let royalty_amount = amount * numerator / denominator;
+            (payee_address, royalty_amount)
+        } else {
+            let royalty = tokenv2::royalty(listing.object);
+            if (option::is_some(&royalty)) {
+                let royalty = option::destroy_some(royalty);
+                let payee_address = royalty::payee_address(&royalty);
+                let royalty_amount =
```

Comment:
You can use https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-stdlib/sources/math64.move#L32